README.md
#naija_emojis

[![Build Status](https://travis-ci.org/andela-womokoro/naija_emojis.svg)](https://travis-ci.org/andela-womokoro/naija_emojis)
[![Coverage Status](https://coveralls.io/repos/andela-womokoro/naija_emojis/badge.svg?branch=master&service=github)](https://coveralls.io/github/andela-womokoro/naija_emojis?branch=master)


This is a simple RESTFul API rendering emoji service to users. It was built with PHP and implements the Slim framework and JSON web token (JWT) for user authentication.

## Install

Via Composer

``` bash
$ composer require Wilson/naija-emoji
```

## Usage

- Create a new users

```
Send a POST request to https://w-naija-emoji.herokuapp.com/register
request body:
{
    'username'  : 'wil',
    'password'  : '******'
    'name'      : 'Wilson Omokoro'
}

response
{
    'status'    : '200'
    'Message'   : 'User registration successful.'
}
```

- User login

```
Send a POST request to https://w-naija-emoji.herokuapp.com/auth/login
request body:
{
    'username'  : 'wil',
    'password'  : '******'
}

A JSON Web Token should be returned on successful login

response
{
    'status'    : '200'
    'Message'   : 'eyJ0eXAiOiJKV1QiLCJh.bGciOiJIUzI1NiJ9.eyJpYXQiOjE0NT'
}
```

- User logout

```
Send a GET request to https://w-naija-emoji.herokuapp.com/auth/logout

request header
{
    'Authorization' : 'eyJ0eXAiOiJKV1QiLCJh.bGciOiJIUzI1NiJ9.eyJpYXQiOjE0NT';
}
```

- Create an Emoji

```
Send a POST request to https://w-naija-emoji.herokuapp.com/emojis
request body:
{
    'name'  : 'Happy'
    'emoji_char'  : 😀
    'category'  : 'Facial'
    'key_words'  : 'happy, glad, delighted'
    'created_by'  : 'Wilson Omokoro'
}
request header
{
    'Authorization' : 'eyJ0eXAiOiJKV1QiLCJh.bGciOiJIUzI1NiJ9.eyJpYXQiOjE0NT';
}

response
{
    'status'    : '201'
    'Message'   : 'Emoji creation successful.'
}
```

- Show all Emojis in the database

```
Send a GET request to https://w-naija-emoji.herokuapp.com/emojis

Response should contain all emojis in JSON format.
```

- Get a particular Emoji

```
Send a GET request containing the Emoji's position in the emojis table to https://w-naija-emoji.herokuapp.com/emojis

E.g. to get the third emoji:

https://w-naija-emojis.herokuapp.com/emojis/3

Response should contain the emoji in JSON format.
```

- Partialy update an Emoji

```
Send a PATCH request containing the Emoji's position in the emojis table to https://w-naija-emoji.herokuapp.com/emojis

E.g. to update just the name of the first emoji:

https://w-naija-emojis.herokuapp.com/emojis/1

request body
{
    'name'  : 'some other name',
}
request header
{
    'Authorization' : 'eyJ0eXAiOiJKV1QiLCJh.bGciOiJIUzI1NiJ9.eyJpYXQiOjE0NT';
}

response
{
    'status'    : '200'
    'Message'   : 'Emoji successfully updated.'
}
```

- Fully update an Emoji

```
Send a PUT request containing the Emoji's position in the emojis table to https://w-naija-emoji.herokuapp.com/emojis

E.g. to fully update the first emoji:

https://w-naija-emojis.herokuapp.com/emojis/1

request body
{
    'name'  : 'Happy'
    'emoji_char'  : 😀
    'category'  : 'Facial'
    'key_words'  : 'happy, glad, delighted'
    'created_by'  : 'Wilson Omokoro'
}
request header
{
    'Authorization' : 'eyJ0eXAiOiJKV1QiLCJh.bGciOiJIUzI1NiJ9.eyJpYXQiOjE0NT';
}

response
{
'status'    : '200'
    'Message'   : 'Emoji successfully updated.'
}
```

- Delete a particular Emoji

```
Send a DELETE request containing the Emoji's position in the emojis table to https://w-naija-emoji.herokuapp.com/emojis

E.g. to DELETE the third emoji:

https://w-naija-emojis.herokuapp.com/emojis/3

request header
{
    'Authorization' : 'eyJ0eXAiOiJKV1QiLCJh.bGciOiJIUzI1NiJ9.eyJpYXQiOjE0NT';
}

response
{
    'status'    : '200'
    'Message'   : 'Emoji 3 deletion successful.'
}
```


## Testing

If the  folder containing your test classes is "tests"

```
$ phpunit tests
```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CONDUCT](CONDUCT.md) for details.

## Security

If you discover any security related issues, please email wilson.omokoro@andela.com instead of using the issue tracker.

## Credits

Naija-emoji is maintained by Wilson Omokoro.

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/league/:package_name.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/thephpleague/:package_name/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/thephpleague/:package_name.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/thephpleague/:package_name.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/league/:package_name.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/league/:package_name
[link-travis]: https://travis-ci.org/thephpleague/:package_name
[link-scrutinizer]: https://scrutinizer-ci.com/g/thephpleague/:package_name/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/thephpleague/:package_name
[link-downloads]: https://packagist.org/packages/league/:package_name
[link-author]: https://github.com/:author_username
[link-contributors]: ../../contributors

CHANGELOG.md
# Changelog

All Notable changes to `naija-emojis` will be documented in this file

## NEXT - YYYY-MM-DD

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


src/controllers/EmojiController.php
<?php

namespace Wilson\Source\Controllers;

use Exception;
use Slim\Slim;
use Wilson\Source\Models\Emoji;
use Wilson\Source\Authenticator;

class EmojiController
{
    /**
     * This method creates an instance of an emoji and stores it in the database.
     * @param  Slim   $app      [Slim instance]
     */
    public static function createEmoji(Slim $app)
    {
        $app->response->headers->set('Content-Type', 'application/json');

        try {
            $auth = Authenticator::authenticate($app);
            $auth = json_decode($auth);

            if(is_object($auth)) {

                $emoji = new Emoji();

                $emoji->name = $app->request->params('name');
                $emoji->emoji_char = $app->request->params('emoji_char');
                $emoji->category = $app->request->params('category');
                $emoji->keywords = $app->request->params('keywords');
                $emoji->created_by = $app->request->params('created_by');

                $rows = $emoji->save();

                if($rows > 0) {
                    $app->halt(201, json_encode("Emoji creation successful."));
                }
                else {
                    throw new Exception("Emoji creation failed!");
                }
            }
            else {
                $app->halt(401, json_encode("You're not authorized to perform this action. Please login."));
            }
        }
        catch(Exception $e) {
            return json_encode($e->getMessage());
        }
    }

    /**
     * This method fetches all the emojis in the database
     * @param  Slim   $app      [Slim instance]
     * @return json encoded string
     */
    public static function getAllEmojis(Slim $app)
    {
        $app->response->headers->set('Content-Type', 'application/json');

        $returnedValue = Emoji::getAll();

        if(is_array($returnedValue)) {
            return json_encode($returnedValue);
        }

        if(is_string($returnedValue)) {
            $app->halt(204, json_encode($returnedValue));
        }
    }

    /**
     * This method is used to search for a particular emoji
     * @param  Slim   $app      [Slim instance]
     * @param  $id              [the id of the emoji in the emojis table]
     * @return json encoded string
     */
    public static function findEmoji(Slim $app, $id)
    {
        $app->response->headers->set('Content-Type', 'application/json');

        $returnedValue = Emoji::findById($id);

        if(is_object($returnedValue)) {
            return $returnedValue->result;
        }

        if(is_string($returnedValue)) {
            $app->halt(404, json_encode($returnedValue));
        }
    }

    /**
     * This function fully updates an emoji in the database i.e. updates all fields of an emoji record
     * @param  Slim   $app      [Slim instance]
     * @param  $id              [the id of the emoji in the emojis table]
     */
    public static function updateEmoji(Slim $app, $id)
    {
        $app->response->headers->set('Content-Type', 'application/json');

        $auth = Authenticator::authenticate($app);
        $auth = json_decode($auth);

        if(is_object($auth))
        {
            $emoji = Emoji::findById($id);

            if(is_string($emoji)) {
                $app->halt(404, json_encode($emoji));
            }

            $emoji->name = self::checkParamValue($app, "name", $app->request->params('name'));
            $emoji->emoji_char = self::checkParamValue($app, "emoji_char", $app->request->params('emoji_char'));
            $emoji->category = self::checkParamValue($app, "category", $app->request->params('category'));
            $emoji->keywords = self::checkParamValue($app, "keywords", $app->request->params('keywords'));
            $emoji->created_by = self::checkParamValue($app, "created_by", $app->request->params('created_by'));

            $rows = $emoji->save();

            if($rows > 0) {
                $app->halt(200, json_encode("Emoji successfully updated."));
            }
            else {
                $app->halt(401, json_encode("Emoji full update failed!"));
            }
        }
    }

    /**
     * This function checks if a parameter contained in a request has a valid value
     * @param  Slim   $app   [Slim instance]
     * @param  $param [the parameter to check]
     * @return string
     */
    public static function checkParamValue(Slim $app, $param, $value)
    {
        $app->response->headers->set('Content-Type', 'application/json');
        echo "value: $value";

        if(is_null($value) || empty($value)) {
           $app->halt(401, json_encode("Cannot update. Missing or invalid parameter: $param"));
        }
        else {
            return $value;
        }
    }

    /**
     * This function updates some fields of an empji record in the database
     * @param  Slim   $app      [Slim instance]
     * @param  $id              [the id of the emoji in the emojis table]
     */
    public static function patchEmoji(Slim $app, $id)
    {
        $app->response->headers->set('Content-Type', 'application/json');

        $auth = Authenticator::authenticate($app);
        $auth = json_decode($auth);

        if(is_object($auth))
        {
            try {
                $emoji = Emoji::findById($id);

                if(is_string($emoji)) {
                    $app->halt(404, json_encode($emoji));
                }

                $params = $app->request->patch();

                foreach ($params as $key => $value) {
                    $emoji->$key = $value;
                }

                $rows = $emoji->save();

                if($rows > 0) {
                    $app->halt(200, json_encode("Emoji successfully updated."));
                }
                else {
                    throw new Exception("Emoji partial update failed!");
                }
            }
            catch(Exception $e) {
                return json_encode($e->getMessage());
            }
        }
    }

    /**
     * This method is used to delete an emoji from the database
     * @param  Slim   $app      [Slim instance]
     * @param  $id              [the id of the emoji in the emojis table]
     * @return json encoded string
     */
    public static function deleteEmoji(Slim $app, $id)
    {
        $app->response->headers->set('Content-Type', 'application/json');

        $auth = Authenticator::authenticate($app);
        $auth = json_decode($auth);

        if(is_object($auth)) {
            $rows = Emoji::destroyById($id);

            if(is_string($rows)) {
                $app->halt(404, json_encode($rows));
            }

            if($rows > 0) {
                return json_encode("Emoji with ID $id deleted successfully.");
            }
            else {
                return json_encode("Failed to delete Emoji with ID $id.");
            }
        }
        else {
            $app->halt(401, json_encode("You're not authorized to perform this action. Please login."));
        }
    }
}